src/Cones/epitrrelentropytri.jl

```diff
@@ -440,85 +455,109 @@ function Δ3!(Δ3::Array{T, 3}, Δ2::Matrix{T}, λ::Vector{T}) where {T <: Real}
     return Δ3
 end
 
+# TODO consider refactor with eig_dot_kron!
 function d2zdV2!(
     d2zdV2::Matrix{T},
-    vecs::Matrix{T},
-    inner::Matrix{T},
+    vecs::Matrix{R},
+    inner::Matrix{R},
     Δ3::Array{T, 3},
-    ten3d::Array{T, 3}, # temp
-    matd2::Matrix{T}, # temp
-    mat::Matrix{T}, # temp
-    mat2::Matrix{T}, # temp
+    ten3d::Array{R, 3}, # temp
+    mat::Matrix{R}, # temp
+    mat2::Matrix{R}, # temp
+    mat3::Matrix{R}, # temp
     rt2::T,
-) where {T <: Real}
+) where {T <: Real, R <: RealOrComplex{T}}
     d = size(vecs, 1)
-
-    @inbounds for i in 1:d
-        @views ten3d_i = ten3d[i, :, :]
-        @views Δ3_i = Δ3[i, :, :]
-        @. mat2 = inner * Δ3_i
-        spectral_outer!(ten3d_i, vecs, Symmetric(mat2, :U), mat)
-    end
-    @inbounds for j in 1:d, i in 1:j
-        @views temp_ij = matd2[block_idxs(d, i), block_idxs(d, j)]
-        @views D_ij = ten3d[:, i, j]
-        spectral_outer!(temp_ij, vecs, D_ij, mat)
+    V = copyto!(mat, vecs')
+    V_views = [view(V, :, i) for i in 1:d]
+    rt2i = inv(rt2)
+    scals = (R <: Complex{T} ? [rt2i, rt2i * im] : [rt2i])
+    @inbounds for k in 1:d
+        @. @views ten3d[:, :, k] = inner * Δ3[:, :, k]
     end
-    copytri!(matd2, 'U')
 
-    rt2i = inv(rt2)
-    row_idx = 1
-    @inbounds for j in 1:d, i in 1:j
-        col_idx = 1
-        ijeq = (i == j)
-        di = d * (i - 1)
-        dj = d * (j - 1)
-        for l in 1:d, k in 1:l
-            rho = (k == l ? (ijeq ? T(0.5) : rt2i) : (ijeq ? rt2i : one(T)))
-            dlk = d * (l - 1) + k
-            dkl = d * (k - 1) + l
-            dji = dj + i
-            dij = di + j
-            d2zdV2[row_idx, col_idx] =
-                rho *
-                (matd2[dji, dlk] + matd2[dij, dlk] + matd2[dji, dkl] + matd2[dij, dkl])
+    col_idx = 1
+    @inbounds for j in 1:d
+        for i in 1:(j - 1), scal in scals
+            mul!(mat3, V_views[j], V_views[i]', scal, false)
+            @. mat2 = mat3 + mat3'
+            for k in 1:d
+                @views mul!(mat3[:, k], ten3d[:, :, k], mat2[:, k])
+            end
+            # mat2 = vecs * (mat3 + mat3) * vecs'
+            @. mat2 = mat3 + mat3'
+            mul!(mat3, Hermitian(mat2, :U), V)
+            mul!(mat2, V', mat3)
+            @views smat_to_svec!(d2zdV2[:, col_idx], mat2, rt2)
             col_idx += 1
         end
-        row_idx += 1
-    end
 
+        mul!(mat2, V_views[j], V_views[j]')
+        for k in 1:d
+            @views mul!(mat3[:, k], ten3d[:, :, k], mat2[:, k])
+        end
+        @. mat2 = mat3 + mat3'
+        mul!(mat3, Hermitian(mat2, :U), V)
+        mul!(mat2, V', mat3)
+        @views smat_to_svec!(d2zdV2[:, col_idx], mat2, rt2)
+        col_idx += 1
+    end
     return d2zdV2
 end
 
+# NOTE helper for hess_prod, totally untested, can have typos
+# function d2zdV2_prod!(
```

Comment:
@lkapelevich @araujoms what should we do with this commented-out hessian product related code? might be worth trying to get it working?

Reply:
ah thanks and good point. OK i'll see if i can add a native test

Reply:
added native tests for the complex case

Reply:
thanks, just added your code in the latest commit.